### PR TITLE
Adding support for AWS T3 instance family

### PR DESCRIPTION
**Pull request by jeli8-cor**
[PLATENG-741]

[PLATENG-741]: https://coralogix.atlassian.net/browse/PLATENG-741

### CODE REVIEW

File: ec2/InstanceType.dhall [hunk 1]
```diff
@@ -3,7 +3,14 @@ let imports = ../imports.dhall
...
 
 let InstanceType =
-      < c5-large
+      < t3-nano
+      | t3-micro
+      | t3-small
+      | t3-medium
+      | t3-large
+      | t3-xlarge
+      | t3-2xlarge
```
Comment (ari-becker) [inline ec2/InstanceType.dhall:12]: This list was maintained in alphabetical order for the instance family - please move t3 to the end

File: ec2/InstanceType.dhall [hunk 2]
```diff
@@ -543,12 +538,19 @@ let alternatives =
...
                           )
+                , t3-nano = [ InstanceType.c5-large, InstanceType.c5d-large ]
+                , t3-micro = [ InstanceType.c5-large, InstanceType.c5d-large ]
+                , t3-small = [ InstanceType.c5-large, InstanceType.c5d-large ]
+                , t3-medium = [ InstanceType.c5-large, InstanceType.c5d-large ]
+                , t3-large = [ InstanceType.c5-large, InstanceType.c5d-large ]
+                , t3-xlarge =
+                  [ InstanceType.c5-xlarge, InstanceType.c5d-xlarge ]
+                , t3-2xlarge =
+                  [ InstanceType.c5-2xlarge, InstanceType.c5d-2xlarge ]
```
Comment (ari-becker) [inline ec2/InstanceType.dhall:553]: Not the correct set of alternatives.
1) for example a t3.large has 8 GB RAM whereas a c5.large has 4 GB RAM, not as much. Same issue for t3.xlarge and t3.2xlarge
2) you need to check for `_params.permitCrossInstanceFamily` before switching from `t*` to `c*`
3) Ignores `m*`, `r*` options

File: ec2/InstanceType.dhall [hunk 3]
```diff
@@ -1328,13 +1196,20 @@ let alternatives =
...
-          { c5-large = Some InstanceType.c5-xlarge
+    = λ(instanceType : InstanceType) →
+        merge
+          { t3-nano = Some InstanceType.t3-nano
+          , t3-micro = Some InstanceType.t3-micro
+          , t3-small = Some InstanceType.t3-small
+          , t3-medium = Some InstanceType.t3-medium
+          , t3-large = Some InstanceType.t3-large
+          , t3-xlarge = Some InstanceType.t3-xlarge
+          , t3-2xlarge = Some InstanceType.t3-2xlarge
```
Comment (ari-becker) [inline ec2/InstanceType.dhall:1211]: You didn't upsize anything here, you returned the same size?

File: ec2/InstanceType.dhall [hunk 4]
```diff
@@ -1390,10 +1265,17 @@ let upsize
...
+    , upsize
+    , specs
+    , alternatives
+    , t3-nano = Some InstanceType.t3-nano
+    , t3-micro = Some InstanceType.t3-micro
+    , t3-small = Some InstanceType.t3-small
+    , t3-medium = Some InstanceType.t3-medium
+    , t3-large = Some InstanceType.t3-large
+    , t3-xlarge = Some InstanceType.t3-xlarge
+    , t3-2xlarge = Some InstanceType.t3-2xlarge
```
Comment (ari-becker) [inline ec2/InstanceType.dhall:1278]: Should return `InstanceType` not `Some InstanceType`

File: ec2/InstanceType.dhall (hunk 3 above)
Comment (jeli8-cor) [reply ec2/InstanceType.dhall:1211]: fixed

File: ec2/InstanceType.dhall (hunk 2 above)
Comment (jeli8-cor) [reply ec2/InstanceType.dhall:553]: 1 - yea, didn't understand the test and what it enforcing. fixed.
2 - I'll add a check
3 - do you mean to add support in these types?

File: ec2/InstanceType.dhall (hunk 3 above)
Comment (ari-becker) [reply ec2/InstanceType.dhall:1211]: Not fixed

Comment (jeli8-cor) [reply ec2/InstanceType.dhall:1211]: didn't push yet. wait for it! :)

File: ec2/InstanceType.dhall (hunk 2 above)
Comment (ari-becker) [reply ec2/InstanceType.dhall:553]: No I mean, if a `t3.large` has 2 vCPU and 8 GB RAM then an `m5.large`, which also has 2 vCPU and 8 GB RAM, should be an alternative for the `t3.large`.

File: ec2/InstanceType.dhall (hunk 3 above)
Comment (ari-becker) [reply ec2/InstanceType.dhall:1211]: Please push then comment ;)

File: ec2/InstanceType.dhall [hunk 5]
```diff
@@ -1267,40 +1137,102 @@ let alternatives =
...
+                    if    _params.permitWorseNetwork
                     then  [ InstanceType.r5-4xlarge, InstanceType.r5a-4xlarge ]
-
+                    else  [] : List InstanceType
+                , t3-nano =
+                    if    _params.permitCrossInstanceFamily
+                    then  [ InstanceType.c5-large
+                          , InstanceType.c5d-large
+                          , InstanceType.m5-large
+                          , InstanceType.m5d-large
```
Comment (ari-becker) [inline ec2/InstanceType.dhall:1160]: There are also `m5n` and `m5dn` which are acceptable substitutes (yes, sometimes they are cheaper on the spot market than an `m5`, it was an interesting surprise we discovered when we first wrote this).

You should also check for the `_params.permitLowerClassCPU` and, if enabled, permit the following families: `m4`, `m5a`, `m5ad`, and `r5a`.

Comment (jeli8-cor) [reply ec2/InstanceType.dhall:1160]: I'll add m5n, m5dn. As for the lower class cpu --> for me, I need it strict and can't allow this. If we'll need it, I'll add it in later phase

**Pull request merged**